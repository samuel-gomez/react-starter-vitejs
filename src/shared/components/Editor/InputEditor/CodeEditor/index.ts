export { default, useCodeEditor, getTemplate } from './CodeEditor';
export type { TReturnUseCodeEditor } from './CodeEditor';
