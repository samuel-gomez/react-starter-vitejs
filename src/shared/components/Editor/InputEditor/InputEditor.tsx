import { CheckboxItem, DateInput, Select, Text } from '@axa-fr/react-toolkit-all';
import type { ComponentPropsWithoutRef, OptionHTMLAttributes } from 'react';
import { DEFAULT_OPTION_LABEL } from 'shared/constants';
import type { TonChange } from '../Editor';
import CodeEditor from './CodeEditor';
import type { TlistElements } from './CodeEditor/Templates';
import JsonEditor from './JsonEditor';

type TListSelect = {
  value: string;
  options?: Record<string, string>;
  type?: string;
  list?: TlistElements;
  labelBtnOpenCodeEditor?: string;
};

export type TInputEditor = {
  value: string | boolean | TListSelect;
  name: string;
  onChange: TonChange;
  options?: { label: string; value: string }[];
};

const commonProps = ({ onChange, name }: Omit<TInputEditor, 'value'>) => ({
  onChange,
  name,
  id: name,
});

type TInputEditorObject = Omit<TInputEditor, 'value'> & {
  value: TListSelect;
};

const InputEditorObject = ({ value, ...props }: TInputEditorObject) => (
  <>
    {(() => {
      switch (true) {
        case !!value.type && value.type === 'date':
          return (
            <DateInput
              crossOrigin={undefined}
              {...commonProps(props)}
              label={props.name}
              onChange={props.onChange as unknown as ComponentPropsWithoutRef<typeof DateInput>['onChange']}
              classModifier="date-custom"
              value={new Date(value.value)}
              classNameContainerLabel="col-md-0"
              classNameContainerInput="col-md-12"
            />
          );
        case !!value.type && value.type === 'separator':
          return <h3 className="af-form-editor__separator">{value.value}</h3>;
        case !!value.type && value.type === 'jsx':
          return <CodeEditor {...commonProps(props)} labelBtnOpenCodeEditor={value.labelBtnOpenCodeEditor} list={value?.list} value={value.value} />;
        case !!value.type && value.type === 'json':
          return <JsonEditor {...commonProps(props)} labelBtnOpenCodeEditor={value.labelBtnOpenCodeEditor} value={value.value} />;
        default:
          return (
            <div className="af-form__select">
              <Select
                {...commonProps(props)}
                forceDisplayPlaceholder
                placeholder={DEFAULT_OPTION_LABEL}
                options={value.options as unknown as OptionHTMLAttributes<HTMLOptionElement>[]}
                value={value.value}
                aria-label={`select-${props.name}`}
              />
            </div>
          );
      }
    })()}
  </>
);

const InputEditor = ({ value, ...props }: TInputEditor) => (
  <>
    {(() => {
      switch (true) {
        case typeof value === 'boolean':
          return (
            <CheckboxItem {...commonProps(props)} disabled={false} isChecked={!!value} value={`${value}`} className="af-form__checkbox-toggle" />
          );

        case typeof value === 'object':
          return <InputEditorObject value={value} {...props} />;

        default:
          return <Text {...commonProps(props)} type="text" value={`${value}`} />;
      }
    })()}
  </>
);

export default InputEditor;
