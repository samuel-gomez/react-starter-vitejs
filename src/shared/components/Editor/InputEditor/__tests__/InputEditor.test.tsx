import { render } from '@testing-library/react';
import InputEditor from '../InputEditor';

describe('InputEditor', () => {
  const onChange = vi.fn();
  it.each`
    value
    ${true}
    ${'value'}
    ${{ value: 'value', options: [] }}
    ${{ value: new Date('2022-12-06'), type: 'date' }}
    ${{ value: 'value', type: 'jsx', labelBtnOpenCodeEditor: 'Edit JSX' }}
    ${{ value: 'value', type: 'json', labelBtnOpenCodeEditor: 'Edit Json' }}
    ${{ value: 'value', type: 'separator' }}
  `('Should render when value: $value', ({ value }) => {
    const { asFragment } = render(<InputEditor value={value} name="name" onChange={onChange} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
