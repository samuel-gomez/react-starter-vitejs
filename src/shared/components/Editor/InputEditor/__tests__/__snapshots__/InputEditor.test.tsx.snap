// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputEditor > Should render when value: 'value' 1`] = `
<DocumentFragment>
  <input
    class="af-form__input-text"
    id="name"
    name="name"
    type="text"
    value="value"
  />
</DocumentFragment>
`;

exports[`InputEditor > Should render when value: { value: 'value', options: [] } 1`] = `
<DocumentFragment>
  <div
    class="af-form__select"
  >
    <div
      class="af-form__select-container"
    >
      <select
        aria-label="select-name"
        class="af-form__input-select"
        id="name"
        name="name"
      >
        <option
          value=""
        >
          - Sélectionner -
        </option>
      </select>
      <span
        aria-controls="name"
        class="glyphicon glyphicon-menu-down"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`InputEditor > Should render when value: { value: 'value', type: 'json', …(1) } 1`] = `
<DocumentFragment>
  <button
    class="btn af-btn af-btn--editor"
    type="submit"
  >
    <span
      class="af-btn__text"
    >
      Edit Json
    </span>
  </button>
</DocumentFragment>
`;

exports[`InputEditor > Should render when value: { value: 'value', type: 'jsx', …(1) } 1`] = `
<DocumentFragment>
  <button
    class="btn af-btn af-btn--editor"
    type="submit"
  >
    <span
      class="af-btn__text"
    >
      Edit JSX
    </span>
  </button>
</DocumentFragment>
`;

exports[`InputEditor > Should render when value: { value: 'value', type: 'separator' } 1`] = `
<DocumentFragment>
  <h3
    class="af-form-editor__separator"
  >
    value
  </h3>
</DocumentFragment>
`;

exports[`InputEditor > Should render when value: { …(2) } 1`] = `
<DocumentFragment>
  <div
    class="row af-form__group af-form__group--date-custom"
  >
    <div
      class="col-md-0"
    >
      <label
        class="af-form__group-label"
        for="name"
      >
        name
      </label>
    </div>
    <div
      class="col-md-12"
    >
      <div
        class="af-form__date"
      >
        <input
          class="af-form__input-date af-form__input-date--date-custom"
          id="name"
          name="name"
          type="date"
          value="2022-12-06"
        />
      </div>
      <small
        class="af-form__help"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`InputEditor > Should render when value: true 1`] = `
<DocumentFragment>
  <div
    class="af-form__checkbox-toggle"
  >
    <input
      checked=""
      class="af-form__input-checkbox"
      id="name"
      name="name"
      type="checkbox"
      value="true"
    />
    <label
      class="af-form__label"
      for="name"
    >
      <span
        class="af-form__indicator"
      >
        <i
          class="glyphicon glyphicon-ok"
        />
      </span>
      <span
        class="af-form__description"
      />
    </label>
  </div>
</DocumentFragment>
`;
